sdk/daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala

```diff
@@ -180,18 +212,248 @@ object Hash {
+    */
+  private sealed abstract class NodeBuilder(
+      version: NodeHashVersion,
+      messageDigest: MessageDigestWithContext,
+      valueBuilder: ValueHashBuilder,
+  ) extends TransactionBuilder(messageDigest, valueBuilder) {
+    {
+      messageDigest.update(
+        version.id,
+        contextO(s"${String.format("%02X", version.id)} (node_version)"),
+      )
+    }
+
+    // We pass the message digest from the node to the value builder, this way we can compute a single hash for the entire
+    // node without having to individually hash every value in the node.
+    protected def valueBuilder: ValueHashBuilder
+
+    protected def addWithValueBuilder[T](
+        value: T,
+        valueBuilderFn: ValueHashBuilder => T => ValueHashBuilder,
+    ): this.type = {
+      discard(valueBuilderFn(valueBuilder)(value))
+      this
+    }
+
+    def addNode(node: Node, nodes: Map[NodeId, Node]): NodeBuilder
+  }
+
+  private object NodeBuilder {
+    private[lf] val LanguageVersionToNodeBuilder
+        : Map[LanguageVersion, (MessageDigestWithContext, ValueHashBuilder) => NodeBuilder] = Map(
+      LanguageVersion.v2_1 -> ((messageDigest, valueBuilder) =>
+        new NodeBuilderV1(messageDigest, valueBuilder),
+      )
+    )
+
+    @throws[NodeHashingError]
+    private[lf] def builderForVersion(
+        version: LanguageVersion,
+        messageDigest: MessageDigestWithContext,
+        valueBuilder: ValueHashBuilder,
+    ): NodeBuilder =
+      NodeBuilder.LanguageVersionToNodeBuilder
+        .getOrElse(version, throw NodeHashingError.UnsupportedLanguageVersion(version))
+        .apply(messageDigest, valueBuilder)
+
+    private[lf] def builderForNode(
+        node: Node,
+        messageDigest: MessageDigestWithContext,
+        valueBuilder: ValueHashBuilder,
+    ): NodeBuilder = node.optVersion match {
+      case Some(version) => builderForVersion(version, messageDigest, valueBuilder)
+      // If the node is version agnostic, the builder doesn't matter so pick V1
+      // This only applies to rollback nodes
+      case None => new NodeBuilderV1(messageDigest, valueBuilder)
+    }
   }
 
-  private final class HashMacBuilder(key: Hash) extends Builder {
+  private final class NodeBuilderV1(
+      messageDigest: MessageDigestWithContext,
+      override protected val valueBuilder: ValueHashBuilder,
+  ) extends NodeBuilder(NodeHashVersion.V1, messageDigest, valueBuilder) {
+    private def addCreateNode(create: Node.Create): this.type = {
+      addContext("Create Node")
+        .withContext("Contract Id")(_.addWithValueBuilder(create.coid, _.addCid))
+        .withContext("Package Name")(_.addString(create.packageName))
+        .withContext("Template Id")(_.addIdentifier(create.templateId))
+        .withContext("Arg")(_.addWithValueBuilder(create.arg, _.addTypedValue))
+        .withContext("Signatories")(_.addStringSet(create.signatories))
+        .withContext("Stakeholders")(_.addStringSet(create.stakeholders))
```

Comment:
What about contract keys? Will we write a new hash function for the LF version that will eventually support non-unique contract keys?